File: C++/subarray-sum-equals-k.cpp
// Time:  O(n)
// Space: O(n)

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        auto result{0};
        auto accumulated_sum{0};
        unordered_map<int, int> lookup;
        ++lookup[0];
        for (const auto& num : nums) {
            accumulated_sum += num;
            result += lookup[accumulated_sum - k];
            ++lookup[accumulated_sum];
        }
        return result;
    }
};

File: Python/subarray-sum-equals-k.py
# Time:  O(n)
# Space: O(n)

# Given an array of integers and an integer k,
# you need to find the total number of continuous subarrays whose sum equals to k.
#
# Example 1:
# Input:nums = [1,1,1], k = 2
# Output: 2
#
# Note:
# The length of the array is in range [1, 20,000].
# The range of numbers in the array is [-1000, 1000] and the range of the integer k is [-1e7, 1e7].

class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        result = 0
        accumulated_sum = 0
        lookup = collections.defaultdict(int)
        lookup[0] += 1
        for num in nums:
            accumulated_sum += num
            result += lookup[accumulated_sum - k]
            lookup[accumulated_sum] += 1
        return result
